flower-via-docker-compose example

NikosVlachakis opened the pull request:
<!--
Thank you for opening a pull request (PR)!

Contribution guidelines: https://github.com/adap/flower/blob/main/CONTRIBUTING.md
-->

## Issue

### Description

<!--
Describe the problem addressed by this PR.

Example: The variable name `rnd` could be misinterpreted as an abbreviation of *random*, but it refers to the current server round.
-->

### Related issues/PRs

<!--
Link issues and/or PRs that are related to this PR.

Example: Fixes #123. See also #456 and #789.
-->

## Proposal

### Explanation

<!--
Explain the changes and how they improve the issue described above.

Example: The variable `rnd` was renamed to `server_round` to improve readability.
-->

### Checklist

- [ ] Implement proposed change
- [ ] Write tests
- [ ] Update [documentation](https://flower.dev/docs/writing-documentation.html)
- [ ] Update [changelog](https://github.com/adap/flower/blob/main/doc/source/changelog.rst)
- [ ] Make CI checks pass
- [ ] Ping maintainers on [Slack](https://flower.dev/join-slack/) (channel `#contributions`)

### Any other comments?

<!--
Please be aware that it may take some time until the maintainers can review the PR.
Smaller PRs with good descriptions can be considered much more easily.

If you have an urgent request or question, please use the Flower Slack:

    https://flower.dev/join-slack/ (channel: #contributions)

Thank you for contributing to Flower!
-->

jafermarq commented on examples/flower-via-docker-compose /README.md:
````diff
@@ -0,0 +1,94 @@
+1. **Execute Initialization Script**: 
+   - Run the `docker_init.sh` script to build the Docker images and start the Docker Compose process. Use the following command:
+     ```bash
+     ./docker_init.sh
````
Maybe saying that everything can be stoped by either doing `docker compose down` (one another terminal if `cd`-ed in the same directory) or `ctrl+c` once.

NikosVlachakis replied:
The initial idea was to provide the .sh script so user can run many instructions together, but its not our case now, so I removed it and added in readme.md the use of --build and how you can stop containers

jafermarq reviewed (commented):
Hi @NikosVlachakis,

This is great :)

I left some comments below. In this first review i mostly focused on following the steps in the readme to run the code. The `grafana` dashboard is not populated with charts automatically. I believe the `uid` needs to be fixed both in the dashboard `.json` and the `prometheus` & `grafana` configs. 

It would be ideal if could add a couple of links to some resources (maybe a grafana tutorial in their docs) so people running the example can look into the reference material in case they want to expand how the dashboard looks like or learn more about grafana. Having links to `prometheus` and `cadvisor` docs is also good.

I like how the docker compose file is generated.

NikosVlachakis commented on examples/flower-via-docker-compose /helpers/generate_docker_compose.py:
```diff
@@ -0,0 +1,127 @@
+      context: .
+      dockerfile: Dockerfile
+    command: python server.py --number_of_rounds={number_of_rounds} --total_clients={total_clients}
+    restart: always
```
The initial decision to use restart: always in the Docker Compose configuration was primarily to address situations where a client encountered an OOM error or similar issue, causing it to exit unexpectedly. This setting ensured that such clients would automatically restart, maintaining their participation in the training process. However, this approach led to a new challenge: after completing a training session (say, 10 rounds), the system would immediately initiate a new training cycle, which was not the intended behavior. To better align with our needs, it might be more appropriate to remove the restart: always policy altogether. This change would prevent automatic restarts of the training, allowing us more control over when and how training sessions begin.

NikosVlachakis commented on examples/flower-via-docker-compose /requirements.txt:
```diff
@@ -0,0 +1,10 @@
+flwr
+tensorflow
+psutil
+ping3
+docker
+scipy
+numpy
```
Totally agree, I run them successfully too.

NikosVlachakis commented on examples/flower-via-docker-compose /README.md:
````diff
@@ -0,0 +1,94 @@
+     80c4a599b2a3   prom/prometheus:latest              "/bin/prometheus --c…"   12 minutes ago   Up 5 minutes             0.0.0.0:9090->9090/tcp                           prometheus
+     169880ab80bd   gcr.io/cadvisor/cadvisor:v0.47.0    "/usr/bin/cadvisor -…"   12 minutes ago   Up 5 minutes (healthy)   0.0.0.0:8080->8080/tcp                           cadvisor
+     ```
+
````
Sounds like a really good idea, showing that capability to the user.

NikosVlachakis commented on examples/flower-via-docker-compose /config/provisioning/dashboards/default_dashboard.json:
```diff
@@ -0,0 +1,139 @@
+        "builtIn": 1,
+        "datasource": {
+          "type": "grafana",
+          "uid": "-- Grafana --"
```
I added the uid

NikosVlachakis commented on examples/flower-via-docker-compose /config/provisioning/datasources/prometheus-datasource.yml:
```diff
@@ -0,0 +1,8 @@
+datasources:
+- name: Prometheus
+  type: prometheus
+  access: proxy
```
I added them so lets see if the dashboard appears when you open grafana

jafermarq reviewed (commented):
Hi @NikosVlachakis,

(following our discussion on Slack) I think this example is in great shape. In my last commits I pushed a few formatting changes that should make this PR pass the CI tests. Also you'll see I added a more concise section early in the readme to show how to run the example. After it, the more detailed section is perfect for those that want to learn more about each component in the example or how to modify it.

Some thoughts:
* Should we move the Grafana dashboard to the top of the README?  maybe at the end of the `Introduction` section?
* I see the name of the directory for the example has a space at the end. Could you rename the directory removing it?

NikosVlachakis commented:
> Hi @NikosVlachakis,
> 
> (following our discussion on Slack) I think this example is in great shape. In my last commits I pushed a few formatting changes that should make this PR pass the CI tests. Also you'll see I added a more concise section early in the readme to show how to run the example. After it, the more detailed section is perfect for those that want to learn more about each component in the example or how to modify it.
> 
> Some thoughts:
> 
> * Should we move the Grafana dashboard to the top of the README?  maybe at the end of the `Introduction` section?
> * I see the name of the directory for the example has a space at the end. Could you rename the directory removing it?

Hey @jafermarq

Regarding moving the Grafana dashboard, I think the current place is good because there is a smooth transition as we go down to the problem of device heterogeneity, and we end up by seeing that graph, that actually shows the things that we have described earlier in the previous sectors. If you think its necessary to move it, for sure.

I removed the space at the end.

Thank you

jafermarq commented:
> > Hi @NikosVlachakis,
> > (following our discussion on Slack) I think this example is in great shape. In my last commits I pushed a few formatting changes that should make this PR pass the CI tests. Also you'll see I added a more concise section early in the readme to show how to run the example. After it, the more detailed section is perfect for those that want to learn more about each component in the example or how to modify it.
> > Some thoughts:
> > 
> > * Should we move the Grafana dashboard to the top of the README?  maybe at the end of the `Introduction` section?
> > * I see the name of the directory for the example has a space at the end. Could you rename the directory removing it?
> 
> Hey @jafermarq
> 
> Regarding moving the Grafana dashboard, I think the current place is good because there is a smooth transition as we go down to the problem of device heterogeneity, and we end up by seeing that graph, that actually shows the things that we have described earlier in the previous sectors. If you think its necessary to move it, for sure.
> 
> I removed the space at the end.
> 
> Thank you

Sure 👍 . Let's keep the dashboard where it is. 

Not sure why but the automated formatting for the markdown file (your `README.md`) is not detecting issues but then the CI test say it is not formatted. Probably we'll need to check manually. I see there are quite a few levels of identation (lists with lists etc -- it could be some of that).

NikosVlachakis commented:
> > > Hi @NikosVlachakis,
> > > (following our discussion on Slack) I think this example is in great shape. In my last commits I pushed a few formatting changes that should make this PR pass the CI tests. Also you'll see I added a more concise section early in the readme to show how to run the example. After it, the more detailed section is perfect for those that want to learn more about each component in the example or how to modify it.
> > > Some thoughts:
> > > 
> > > * Should we move the Grafana dashboard to the top of the README?  maybe at the end of the `Introduction` section?
> > > * I see the name of the directory for the example has a space at the end. Could you rename the directory removing it?
> > 
> > 
> > Hey @jafermarq
> > Regarding moving the Grafana dashboard, I think the current place is good because there is a smooth transition as we go down to the problem of device heterogeneity, and we end up by seeing that graph, that actually shows the things that we have described earlier in the previous sectors. If you think its necessary to move it, for sure.
> > I removed the space at the end.
> > Thank you
> 
> Sure 👍 . Let's keep the dashboard where it is.
> 
> Not sure why but the automated formatting for the markdown file (your `README.md`) is not detecting issues but then the CI test say it is not formatted. Probably we'll need to check manually. I see there are quite a few levels of identation (lists with lists etc -- it could be something around those lines).

Yes I see what you are saying,
maybe there is also something related with the HTML tags in the top of the file?

jafermarq commented:
@Robert-Steiner, please take a look at this example when you have some time. It looks completed to me eyes. But maybe you have some suggestions regarding how docker/docker-compose (or even grafana/prometheus) that we should incorporate to make this example more future proof.

jafermarq reviewed (dismissed):
@NikosVlachakis, this looks great!

danieljanes reviewed (commented):
Great PR, looking forward to merging this soon!

Just one small detail: for the images in `public`, could we remove them and link them similar to how we do it with the Flower logo?

NikosVlachakis commented:
> Great PR, looking forward to merging this soon!
> 
> Just one small detail: for the images in `public`, could we remove them and link them similar to how we do it with the Flower logo?

Hey Daniel, thank you.
Do you mean to remove the entire public folder and use external (public) links for each  image?

danieljanes commented:
> Do you mean to remove the entire public folder and use external (public) links for each image?

Yes, exactly. When we link to images, we don't need to commit them to the repo, and it's clear where they come from for users.

NikosVlachakis commented:
> > Do you mean to remove the entire public folder and use external (public) links for each image?
> 
> Yes, exactly. When we link to images, we don't need to commit them to the repo, and it's clear where they come from for users.

Ok, I see, I removed the public folder and created publicly available urls for each image.
